docs/tasks/administer-cluster/setup-ha-etcd-with-kubeadm.md

```diff
@@ -0,0 +1,143 @@
+not a great default for a highly available control plane. This task walks
+through the process of creating a highly available etcd cluster of three members
+that can be used as an external etcd when using kubeadm to set up a kubernetes
+cluster.
+{% endcapture %}
+
+{% capture prerequisites %}
+* Three hosts that can talk to each other over ports 2379 and 2380. This
+  document assumes these default ports. However, they are configurable through
+  the kubeadm config file.
+* Each host must [have docker, kubelet, and kubeadm installed][toolbox].
+* SSH between hosts is required for certificate copying.
+
+[toolbox]: /docs/setup/independent/install-kubeadm/
+{% endcapture %}
+
+{% capture steps %}
+
+## Kubeadm Configuration File
+
+Create a kubeadm configuration file on each host that looks something like this:
+
+Pay attention to the `name` parameter in `extraArgs`. It needs to be unique per
+host. This example uses infra0, infra1, and infra2, but it can be any valid etcd
+member name just so long as the names all match in the `initial-cluster`
+section.
+
+```
+export HOST1=10.0.0.1 # Update HOST1, HOST2, and HOST3 with the IPs or resolvable names of your hosts
+export HOST2=10.0.0.2
+export HOST3=10.0.0.3
+export CURRENT_HOST="${HOST1}" # This must change for each host
+export NAME=infra0 # Update to use infra0, infra1, or infra2
+
+cat << EOF > kubeadmcfg.yaml
+apiVersion: "kubeadm.k8s.io/v1alpha1"
+kind: MasterConfiguration
+etcd:
+    serverCertSANs:
+    - "${CURRENT_HOST}"
+    peerCertSANs:
+    - "${CURRENT_HOST}"
+    extraArgs:
+        initial-cluster-token: etcd-cluster-1
+        initial-cluster: infra0=https://${HOST1}:2380,infra1=https://${HOST2}:2380,infra2=https://${HOST3}:2380
+        initial-cluster-state: new
+        name: ${NAME}
+        listen-peer-urls: https://${CURRENT_HOST}:2380
+        listen-client-urls: https://${CURRENT_HOST}:2379
+        advertise-client-urls: https://${CURRENT_HOST}:2379
+        initial-advertise-peer-urls: https://${CURRENT_HOST}:2380
+EOF
+```
+
+## Certificates
+
+### Certificate Authority
+
+If you already have a CA then the only action that is required is copying the
+`crt` and `key` file to `/etc/kubernetes/pki/etcd/ca.crt` and
+`/etc/kubernetes/pki/etcd/ca.key`. After those files have been copied, please
+skip to the Certificate Swizzling section below.
+
+Create a CA for the etcd certs. This only has to be done on one node.
+
+```
+kubeadm alpha phase certs etcd-ca
+```
+
+This creates two files
+
+1. `/etc/kubernetes/pki/etcd/ca.crt`
+2. `/etc/kubernetes/pki/etcd/ca.key`
+
+### Certificate Swizzling
+
+Copy these certs to the other two nodes and put them in `/etc/kubernetes/pki/etcd/`.
+
+The steps to get these files on `$HOST2` might look like this:
+
+```
+root@HOST1 $ scp /etc/kubernetes/pki/etcd/ca.crt ubuntu@${HOST2}:/home/ubuntu/
+root@HOST1 $ scp /etc/kubernetes/pki/etcd/ca.key ubuntu@${HOST2}:/home/ubuntu/
```

Comment:
Is the ca key needed for running an etcd peer on every master? Otherwise you might want to point out that you can also pregenerate everything and only distribute the unique client/serving certs to each master as well as the ca cert (only)

=> For security-concious folks

Reply:
It seems like this etcd ca key is only required when adding new kube master with etcd running in it. If so I agree that once the ca key is not needed, kubeadmn advise users to remove it from the server and keep the copy in the safe place.

Reply:
I waffled on this bit as well. At first I had it this way, but this required a ton of instructions for moving files around and changing permissions. I ended up going this route to reduce the number of copy instructions. I also received feedback to actually provide a list of commands to move files around, hence the original requirement for SSH between nodes.

Unfortunately for me, I think you're right that I *actually* want to distribute the certs after generating them all on one host. I will have to re-write the workflow.